service-userlayer/src/main/java/org/oskari/map/userlayer/service/UserLayerDbServiceMybatisImpl.java

```diff
@@ -266,74 +277,107 @@ private UserLayerMapper getMapper(SqlSession session) {
         try {
-            json.put(GeoJSON.TYPE, GeoJSON.FEATURE_COLLECTION);
-            json.put("crs", geojsonWriter.writeCRSObject(targetSRSName));
+            if (features == null || features.isEmpty()) {
+                return new EmptyFeatureCollection(null);
+            }
 
-            JSONArray jsonFeatures = new JSONArray(features.stream().map(feature -> this.toGeoJSONFeature(feature, targetSRSName)).collect(Collectors.toList()));
-            json.put(GeoJSON.FEATURES, jsonFeatures);
+            DefaultFeatureCollection collection = new DefaultFeatureCollection();
+            for (UserLayerData feature: features) {
+                collection.add(toSimpleFeature(feature));
+            }
 
-        } catch(JSONException ex) {
-            log.warn("Failed to create GeoJSON FeatureCollection");
-            throw new ServiceException("Failed to create GeoJSON FeatureCollection");
+            return collection;
+        } catch (Exception e) {
+            log.warn(e, "Failed to create SimpleFeatureCollection");
+            throw new ServiceException("Failed to create SimpleFeatureCollection");
         }
-        return json;
     }
 
-     private JSONObject toGeoJSONFeature(UserLayerData feature, String targetSRSName) {
-        JSONObject jsonFeature = new JSONObject();
-        JSONObject properties = new JSONObject();
-        try {
-            jsonFeature.put("id", feature.getId());
-            jsonFeature.put("geometry_name", GeoJSON.GEOMETRY);
-            jsonFeature.put(GeoJSON.TYPE, GeoJSON.FEATURE);
-
-            String sourceSRSName = "EPSG:" + feature.getDatabaseSRID();
-            Geometry transformed = WKTHelper.transform(parseWKT(feature.getWkt()), sourceSRSName, targetSRSName);
-            JSONObject geoJsonGeometry = geojsonWriter.writeGeometry(transformed);
-            jsonFeature.put(GeoJSON.GEOMETRY, geoJsonGeometry);
-
-            properties.put("id", feature.getId());
-            properties.put("user_layer_id", feature.getUser_layer_id());
-            properties.put("uuid", feature.getUuid());
-            properties.put("feature_id", feature.getFeature_id());
-            properties.put("property_json", feature.getProperty_json());
-            properties.put("created", feature.getCreated());
-            properties.put("updated", feature.getUpdated());
-            jsonFeature.put("properties", properties);
-
-        } catch(JSONException ex) {
-            log.warn("Failed to convert UserLayerData to GeoJSONFeature");
+    private SimpleFeature toSimpleFeature(UserLayerData feature) {
+
+        Geometry geom = parseWKT(feature.getWkt());
+        SimpleFeatureTypeBuilder featureTypeBuilder = getFeatureTypeBuilder(geom);
+
+        SimpleFeatureType simpleFeatureType = featureTypeBuilder.buildFeatureType();
+        SimpleFeatureBuilder featureBuilder = new SimpleFeatureBuilder(simpleFeatureType);
+        featureBuilder.set(GEOM_ATTRIBUTE, parseWKT(feature.getWkt()));
```

Comment:
Don't parse WKT again, use the already parsed `geom` instead.

Reply:
You're right about these. Unfortunately it seems that a generic featuretype with Geometry.class does not work. It results in an exception when trying to create the actual feature.

Reply:
Thanks for trying!